sharding/networkSharding/peerShardMapper.go

```diff
@@ -0,0 +1,133 @@
+package networkSharding
+
+import (
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+	"github.com/ElrondNetwork/elrond-go/storage"
+)
+
+// PeerShardMapper stores the mappings between peer IDs and shard IDs
+type PeerShardMapper struct {
+	peerIdPk         storage.Cacher
+	fallbackPkShard  storage.Cacher
+	fallbackPidShard storage.Cacher
+
+	nodesCoordinator sharding.NodesCoordinator
+}
+
+// NewPeerShardMapper creates a new peerShardMapper instance
+func NewPeerShardMapper(
+	peerIdPk storage.Cacher,
+	fallbackPkShard storage.Cacher,
+	fallbackPidShard storage.Cacher,
+	nodesCoordinator sharding.NodesCoordinator,
+) (*PeerShardMapper, error) {
+
+	if check.IfNil(nodesCoordinator) {
+		return nil, sharding.ErrNilNodesCoordinator
+	}
+	if check.IfNil(peerIdPk) {
+		return nil, sharding.ErrNilCacher
+	}
+	if check.IfNil(fallbackPkShard) {
+		return nil, sharding.ErrNilCacher
+	}
+	if check.IfNil(fallbackPidShard) {
+		return nil, sharding.ErrNilCacher
+	}
+
+	return &PeerShardMapper{
+		peerIdPk:         peerIdPk,
+		fallbackPkShard:  fallbackPkShard,
+		fallbackPidShard: fallbackPidShard,
+		nodesCoordinator: nodesCoordinator,
+	}, nil
+}
+
+// ByID returns the corresponding shard ID of a given peer ID.
+// If the pid is unknown, it will return sharding.UnknownShardId
+func (psm *PeerShardMapper) ByID(pid p2p.PeerID) (shardId uint32) {
```

Comment:
I expect to have more ID requests than nodes distribution in stards update , maybe is better to have the ID to shard mapping in one cache, which is maintained by the update calls

Reply:
Ok, we can do that in a next PR. Right now it should suffice and should work. The reason behind this implementation is that I wanted to use an interface (NodesCoordinator) that should be able to efficiently tell the mapper the shard ID for a provided pk. NodesCoordinator is used each time a new header is intercepted because the header interceptor needs to verify signatures with the help of the NodesCoordinator.
Since connecting/disconnecting to/from a peer is a lengthy process (~hundreds of milliseconds) compared to searching through a map of slices with limited number of items, I think we can leave it like this for the moment and improve the updating with an event-based scheme at a later time.